File: js/src/ipc.ts
import { appendFileSync } from "node:fs";
import { v4 as uuid } from "uuid";

import { ChildProcess, spawn } from "node:child_process";
import { EventEmitter } from "node:events";

export interface GoPayload<Data, Error> {
	id: string;
	event: string;
	data: Data;
	error: Error;
	SR: boolean;
}

export type Channel = string | { event: string; id: string };

/**
  Adapted from https://github.com/Akumzy/ipc-node/blob/master/src/index.ts

  MIT License

  Copyright (c) 2019 Akuma Isaac Akuma

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE.
 */
export class IPC extends EventEmitter {
	go: ChildProcess | null;

	closed: boolean;

	constructor(private binPath: string) {
		super();

		this.go = null;

		this.closed = false;

		/**
		 * The `Golang` process will be pinging at every 20 seconds
		 * and will wait another 20 seconds for reply via `pong` event name
		 * else it will kill it's process.
		 */
		this.on("ping", () => this.send("pong"));
	}

	/**
	 * Start the child process
	 * @param arg
	 */
	public init(arg: string[] = []) {
		const self = this;
		const go = spawn(this.binPath, arg, {});

		this.closed = false;
		this.go = go;

		go.stderr.setEncoding("utf8");
		go.stdout.setEncoding("utf8");

		// emit the errors
		go.stderr.on("error", (e) => self.emit("error", e));
		go.stderr.on("data", (e) => self.emit("log", e));

		let outBuffer = "";

		go.stdout.on("data", (s) => {
			outBuffer += s;

			if (s.endsWith("}\\n")) {
				if (process.env.ODICT_DEBUG_IPC) {
					appendFileSync("ipc.log", `NODE RECEIVED: ${outBuffer}`);
				}
				self._processData(outBuffer);
				outBuffer = "";
			}
		});

		go.once("close", (_) => {
			self.closed = true;
			self.emit("close");
		});

		process.on("beforeExit", () => this.kill());

		return this;
	}

	private _processData(payload: string) {
		const _data = this.parseJSON(payload);

		if (Array.isArray(_data)) {
			for (const item of _data) {
				const { id, error, data, event } = item;

				this.emit("data", item);
				this.emit(event, data, error);
				this.emit(`${event}:${id}`, data, error);
			}
		}
	}

	/**
	 * Kill the child process
	 */
	public kill() {
		try {
			this.send("___EXIT___", null);
			this.closed = true;
			this.go?.kill();
			this.emit("killed");
		} catch (error) {
			console.log(error);
		}
	}

	/**
	 * Send message to `Golang` process
	 * @param event
	 * @param data
	 */
	public send<Data>(event: Channel, data: Data | undefined = undefined) {
		this._send(event, data, false);
	}

	/**
	 * sendRaw gives your access to a third `boolean` argument which
	 * is used to determine if this is a sendAndReceive action
	 */
	public sendRaw<Data>(event: Channel, data: Data, isSendAndReceive = false) {
		this._send(event, data, isSendAndReceive);
	}

	/**
	 *
	 * @param event
	 * @param data
	 * @param SR this tells `Go` process if this message needs an acknowledgement
	 */
	private _send<Data>(event: Channel, data: Data, SR: boolean) {
		try {
			if (this.go && !this.closed && this.go.stdin?.writable) {
				const payload =
					typeof data === "object" || Array.isArray(data)
						? JSON.stringify(data)
						: data;

				// We are converting this to `JSON` this to preserve the
				// data types
				const d = JSON.stringify({
					id: typeof event === "string" ? uuid() : event.id,
					event: typeof event === "string" ? event : event.event,
					data: payload,
					SR: !!SR,
				});

				if (this.go.stdin.writable) {
					const text = d + "\n";

					if (process.env.ODICT_DEBUG_IPC) {
						appendFileSync("ipc.log", `NODE SENT: ${text}`);
					}

					this.go.stdin.write(text);
				}
			}
		} catch (error) {
			this.emit("error", error);
		}
	}

	/**
	 *  Send and receive an acknowledgement through
	 * a callback from `Go` process
	 * @param event
	 * @param data
	 * @param cb
	 */
	public sendAndReceive (
		event: string,
		data: any,
		cb: (error: Error, data: any) => void,
	) {
		const id = uuid();

		this._send({ event, id }, data, true);

		const rc = `${event}:${id}`;

		this.once(rc, (data, error) => {
			if (typeof cb === "function") {
				cb(error, data);
			}
		});
	}

	/**
	 *  Receive and send back acknowledgement/data to `GO`
	 * a callback from `Go` process
	 * @param event
	 * @param data
	 * @param cb
	 */
	public onReceiveAnSend<Data>(
		event: string,
		cb: (channel: Channel, data: Data) => void,
	) {
		const channel = { event, id: uuid() };

		this.on(event, (data) => {
			if (typeof cb === "function") {
				cb(channel, data);
			}
		});
	}

	private parseJSON<Data, Error>(s: string): GoPayload<Data,Error>[] | null {
		try {
			let data = s.replace(/}\\n/g, "},");

			if (data.endsWith(",")) {
				data = data.slice(0, -1).trim();
			}

			return JSON.parse(`[${data}]`);
		} catch (error) {
			this.emit("parse-error", error);
			return null;
		}
	}
}

File: cli/ipc.go
/*
Adapted from https://raw.githubusercontent.com/Akumzy/ipc/master/ipc.go

# MIT License

# Copyright (c) 2019 Akuma Isaac Akuma

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
package cli

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/TheOpenDictionary/odict/lib/utils"
	"github.com/google/uuid"
)

// IPC channel
type IPC struct {
	sendChannel           chan payload
	receiveListerners     map[string][]Handler
	receiveSendListerners map[string][]HandlerWithReply
}

var (
	rLock  sync.Mutex
	rRLock sync.Mutex
)

type replyChannel struct {
	Event string `json:"event"`
	ID    string `json:"id"`
}

// Payload this is the payload structure
type payload struct {
	ID    string `json:"id"`
	Event string `json:"event"`
	// If the data received from the parent is a literal value `Data`
	//type will be equals to the underlining type for example:
	// JavaScripts === Go
	// `null  === nil`
	// `undefined === nil`
	// `number(int) === int`
	// `string === string`
	// else if the `Data` is an Object in JavaScript
	// data will be a JSON string
	Data  interface{} `json:"data"`
	Error interface{} `json:"error"`
	SR    bool        `json:"SR"` // send and receive
	RS    bool        `json:"RC"` // receive and send
}

// Handler When the underline type of data is being
//
//	access through `type assertion` if the data has a
//	literal value the underlining type will be return
//	else a `JSON` representative of the data will be return
type Handler func(data interface{})

// HandlerWithReply  When the underline type of data is being
//
//	access through `type assertion` if the data has a literal
//	value the underlining type will be return else a `JSON` representative of
//	the data will be return.
//	`replyChannel` is the event name you'll pass to `ipc.Reply` method to respond
//	 to the sender
type HandlerWithReply func(channel replyChannel, data interface{})

// PayloadReceive this is the payload structure
type payloadReceive struct {
	ID    string      `json:"id"`
	Event string      `json:"event"`
	Data  interface{} `json:"data"`
	SR    bool        `json:"SR"` //send and receive
}

// Send data to parent process
func (ipc IPC) Send(event string, data interface{}) {
	ipc.sendChannel <- payload{ID: uuid.NewString(), Event: event, Data: data}
}

// Reply back to sender
func (ipc IPC) Reply(channel replyChannel, data, err interface{}) {
	ipc.sendChannel <- payload{ID: channel.ID, Event: channel.Event, Data: data, SR: true, Error: err}
}

// On listens for events from parent process
func (ipc IPC) On(event string, handler Handler) {
	rLock.Lock()
	defer rLock.Unlock()
	h := ipc.receiveListerners[event]
	h = append(h, handler)
	ipc.receiveListerners[event] = h
}

// OnReceiveAndReply listen for an events and as well reply back to
// the same sender with the help of `ipc.Reply` method
func (ipc IPC) OnReceiveAndReply(event string, handler HandlerWithReply) {
	rRLock.Lock()
	defer rRLock.Unlock()
	h := ipc.receiveSendListerners[event]
	h = append(h, handler)
	ipc.receiveSendListerners[event] = h
}

// SendAndReceive send and listen for reply event
func (ipc IPC) SendAndReceive(event string, data interface{}, handler Handler) {
	id := uuid.NewString()
	ipc.sendChannel <- payload{ID: id, Event: event, Data: data, RS: true}
	channel := event + ":" + id
	ipc.On(channel, handler)
}

// RemoveListener remove listener
func (ipc IPC) RemoveListener(event string) {
	delete(ipc.receiveListerners, event)
}

// Start `ipc`
// the `ipc.Start` method will blocks executions
// so is either you put in a separate `Go routine` or put you own code in
// a different `Go routine`
func (ipc IPC) Start() {

	go func() {
		for {
			msg := <-ipc.sendChannel
			data, err := Marshal(msg)
			if err != nil {
				log.Println(err)
			} else {
				text := data + "\\n"

				if os.Getenv("ODICT_DEBUG_IPC") == "true" {
					utils.AppendToFile("ipc.log", "GO SENT: "+text)
				}

				fmt.Print(text)
			}
		}
	}()

	reader := bufio.NewReader(os.Stdin)

	for {
		text, err := reader.ReadString('\n')

		if err != nil {
			log.Println(err)
			continue
		}

		if text != "" {
			var payload payloadReceive

			text = strings.TrimSuffix(text, "\n")

			if os.Getenv("ODICT_DEBUG_IPC") == "true" {
				utils.AppendToFile("ipc.log", "GO RECEIVED: "+text)
			}

			// check if the text is not empty string
			if text != "" {
				if err := json.Unmarshal([]byte(text), &payload); err != nil {
					log.Println(err)
					continue
				}

				if payload.Event == "___EXIT___" {
					os.Exit(0)
				}

				// Run the handlers in a goroutine to prevent
				// https://github.com/Akumzy/ipc/issues/1
				go func() {
					if payload.SR {
						for _, handler := range ipc.receiveSendListerners[payload.Event] {
							handler(replyChannel{ID: payload.ID, Event: payload.Event}, payload.Data)
						}
					} else {
						for _, handler := range ipc.receiveListerners[payload.Event+":"+payload.ID] {
							handler(payload.Data)
						}
						println(payload.Event)
						println(payload.SR)
						println(len(ipc.receiveListerners[payload.Event]))
						for _, handler := range ipc.receiveListerners[payload.Event] {
							println(handler)
							handler(payload.Data)
						}
					}
				}()
			}

		}
	}
}

// Marshal to json
func Marshal(v interface{}) (string, error) {
	buf := new(bytes.Buffer)
	enc := json.NewEncoder(buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(&v); err != nil {
		return "", err
	}
	log.Println(buf.String())
	return strings.TrimSpace(buf.String()), nil
}

// pingPong is used to eliminate zombies
// ping the parent process every 20 seconds
func pingPong(ipc *IPC) {
	isActive := true
	ipc.On("pong", func(d interface{}) {
		isActive = true
	})
	for {
		time.Sleep(20 * time.Second)
		if isActive {
			isActive = false
		} else {
			log.Println("[IPC] Timeout closing process.")
			os.Exit(0)
		}
		ipc.Send("ping", nil)
	}
}

// New return now ipc
func NewIPC() *IPC {
	ipc := &IPC{}
	ipc.sendChannel = make(chan payload)
	ipc.receiveListerners = make(map[string][]Handler)
	ipc.receiveSendListerners = make(map[string][]HandlerWithReply)
	go pingPong(ipc)
	return ipc
}
